lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,499 @@
+  }
+
+  return ArrayAttr::get(op->getContext(), newArgAttrs);
+};
+
+void GrandCentralPass::runOnOperation() {
+  CircuitOp circuitOp = getOperation();
+
+  AnnotationSet annotations(circuitOp);
+  if (annotations.empty())
+    return;
+
+  auto context = &getContext();
+  OpBuilder builder(context);
+  builder.setInsertionPointToEnd(getOperation()->getBlock());
+
+  // Store a mapping of interface name to InterfaceOp.
+  llvm::StringMap<sv::InterfaceOp> interfaces;
+
+  // Track the order that interfaces should be emitted in.
+  llvm::SmallVector<std::pair<StringRef, StringRef>> interfaceKeys;
+
+  // Examine the Circuit's Annotations doing work to remove Grand Central
+  // Annotations.  Ignore any unprocesssed annotations and rewrite the Circuit's
+  // Annotations with these when done.
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+  for (auto anno : annotations) {
+    if (!anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType")) {
+      unprocessedAnnos.push_back(anno.getDict());
+      continue;
+    }
+
+    AugmentedBundleType bundle;
+    if (auto maybeBundle = decodeBundleType(anno))
+      bundle = maybeBundle.getValue();
+    else {
+      emitError(circuitOp.getLoc(),
+                "'firrtl.circuit' op contained an 'AugmentedBundleType' "
+                "Annotation which did not conform to the expected format")
+              .attachNote()
+          << "the problematic 'AugmentedBundleType' is: '" << anno.getDict()
+          << "'";
+      return signalPassFailure();
+    }
+
+    for (auto elt : bundle.elements) {
+      std::pair<StringRef, StringRef> key = {bundle.defName, elt.name};
+      interfaceMap[key] = {elt.tpe, elt.description};
+      interfaceKeys.push_back(key);
+    }
+
+    // If the interface already exists, don't create it.
+    if (interfaces.count(bundle.defName))
+      continue;
+
+    // Create the interface.  This will be populated later.
+    interfaces[bundle.defName] =
+        builder.create<sv::InterfaceOp>(circuitOp->getLoc(), bundle.defName);
+  }
+
+  circuitOp->setAttr("annotations", builder.getArrayAttr(unprocessedAnnos));
+
+  // Walk through the circuit to collect additional information.  If this fails,
+  // signal pass failure.
+  for (auto &op : circuitOp.getBody()->getOperations()) {
+    if (isa<FModuleOp, FExtModuleOp>(op)) {
+      GrandCentralVisitor visitor(&getContext(), &interfaceMap);
+      visitor.visitModule(&op);
+      if (visitor.hasFailed())
+        return signalPassFailure();
+    }
+  }
+
+  // Populate interfaces.
+  for (auto a : interfaceKeys) {
+    auto defName = a.first;
+    auto name = a.second;
+
+    auto info = &interfaceMap[{defName, name}];
+    if (info->isMissing()) {
+      emitError(circuitOp.getLoc())
+          << "'firrtl.circuit' op contained a Grand Central Interface with "
+             "defName '"
+          << defName << "' that had a component called '" << name
```

Comment:
```suggestion
      circuitOp.emitOpError(
             "'firrtl.circuit' op contained a Grand Central Interface with "
             "defName '")
          << defName << "' that had a component called '" << name
```
maybe?

Reply:
It was like this before with `emitOpError`. I _intentionally_ switched away from this because `emitOpError` adds a note that includes *the whole op*. This is fine for small circuits, but it's goofy to dump MiB's of circuit. :joy:

I didn't see a way around it, so I'm using a pared down version.

This is entirely non-obvious, so I'll move this to a lambda (as it's used one other place) and add a comment with this information.